File: orchestrator-service/app/monitoring/elk.py
"""
app/monitoring/elk.py
──────────────────────
ELK monitoring backend (Elasticsearch Bulk API + optional Logstash HTTP input).

Mirrors gateway's models/monitoring/elk.go ELKBackend.

Log flow:
  structlog JSON handler
    → multiHandler fan-out
         ├─ original stdout handler  (preserved)
         └─ ELKHandler → buffer → flush worker
              ├─ ES Bulk API       (when logstash is None)
              └─ Logstash HTTP     (when logstash is configured)
"""
from __future__ import annotations

import logging
import threading

from app.monitoring.base import Backend
from app.monitoring._utils import _build_ssl_context
from app.monitoring.config import ELKConfig, ElasticsearchConfig, LogstashConfig

logger = logging.getLogger(__name__)


class ELKBackend(Backend):
    """
    Ships structured log records to Elasticsearch (Bulk API) or via a
    Logstash HTTP input.  Mirrors gateway's models/monitoring/elk.go ELKBackend.
    """

    def __init__(self, cfg: ELKConfig, version: str, commit: str) -> None:
        self._cfg          = cfg
        self._buf: list    = []
        self._buf_lock     = threading.Lock()
        self._flush_event  = threading.Event()
        self._stop_event   = threading.Event()
        self._flush_thread: threading.Thread | None = None
        self._prev_handler = None   # restored on shutdown

        if not cfg.enabled:
            logger.info("elk backend: disabled")
            return

        es_cfg = cfg.elasticsearch
        if not es_cfg or not es_cfg.endpoints:
            logger.warning("elk backend: elasticsearch.endpoints required — disabled")
            return

        # Build ES client (always) and optional Logstash writer.
        self._es = self._build_es_client(es_cfg)
        self._ls = self._build_ls_writer(cfg.logstash) if cfg.logstash else None

        # Install structlog multi-processor (preserves stdout handler).
        self._install_log_handler(cfg)

        # Start background flush worker.
        self._flush_thread = threading.Thread(
            target=self._flush_worker,
            daemon=True,
            name="elk-flush",
        )
        self._flush_thread.start()

        logger.info(
            "elk backend: started service=%s index=%s via_logstash=%s",
            cfg.service_name,
            es_cfg.index,
            cfg.logstash is not None,
        )

    def name(self) -> str:
        return "elk"

    def ping(self) -> None:
        if not hasattr(self, "_es") or self._es is None:
            return
        self._es.ping()
        if self._ls is not None:
            self._ls.ping()

    def shutdown(self) -> None:
        self._stop_event.set()
        if self._flush_thread is not None:
            self._flush_thread.join(timeout=10)
        self._flush()   # final flush
        self._restore_log_handler()
        logger.info("elk backend: shut down")

    # ── enqueue (called by log handler) ───────────────────────────────────────

    def enqueue(self, record: dict) -> None:
        max_docs = self._cfg.elasticsearch.bulk_max_docs or 500
        with self._buf_lock:
            self._buf.append(record)
            should_flush = len(self._buf) >= max_docs
        if should_flush:
            self._flush_event.set()

    # ── flush worker ──────────────────────────────────────────────────────────

    def _flush_worker(self) -> None:
        interval = self._cfg.elasticsearch.flush_interval_seconds or 5.0
        while not self._stop_event.is_set():
            self._flush_event.wait(timeout=interval)
            self._flush_event.clear()
            self._flush()

    def _flush(self) -> None:
        with self._buf_lock:
            if not self._buf:
                return
            batch = self._buf[:]
            self._buf.clear()

        try:
            if self._ls is not None:
                self._ls.write_batch(batch)
            else:
                self._es.bulk_index(batch)
        except Exception:
            logger.warning("elk backend: flush failed", exc_info=True)

    # ── structlog handler installation ────────────────────────────────────────

    def _install_log_handler(self, cfg: ELKConfig) -> None:
        """
        Add a stdlib logging.Handler to the root logger so every log record
        — whether from structlog, FastAPI, uvicorn, Celery or any other
        library that uses logging.getLogger() — is enqueued for ES shipping.

        This mirrors gateway's slog.SetDefault(slog.New(newMultiHandler(...))):
        a second handler is attached alongside the existing stdout handler so
        stdout output is preserved and ELK receives every record.

        Using a stdlib Handler (not a structlog processor) is necessary because
        configure_logging() routes stdlib loggers through ProcessorFormatter's
        foreign_pre_chain, which bypasses the structlog processor chain entirely.
        """
        import datetime as _dt
        import traceback as _tb

        backend   = self
        svc_name  = cfg.service_name
        svc_ver   = cfg.service_version or "dev"
        env_label = cfg.environment
        es_cfg    = cfg.elasticsearch

        class _ELKHandler(logging.Handler):
            def emit(self, record: logging.LogRecord) -> None:
                try:
                    doc: dict = {
                        "@timestamp":      _dt.datetime.utcnow().strftime(
                                               "%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
                        "level":           record.levelname.lower(),
                        "message":         record.getMessage(),
                        "logger":          record.name,
                        "service.name":    svc_name,
                        "service.version": svc_ver,
                        "environment":     env_label,
                    }
                    if es_cfg and es_cfg.extra_fields:
                        doc.update(es_cfg.extra_fields)
                    if record.exc_info:
                        doc["exc_info"] = "".join(_tb.format_exception(*record.exc_info))
                    if record.stack_info:
                        doc["stack_info"] = record.stack_info
                    backend.enqueue(doc)
                except Exception:
                    pass  # never crash the logging subsystem

        self._elk_handler = _ELKHandler(level=logging.DEBUG)
        logging.getLogger().addHandler(self._elk_handler)

    def _restore_log_handler(self) -> None:
        """Remove the ELK stdlib handler from the root logger on shutdown."""
        handler = getattr(self, "_elk_handler", None)
        if handler is not None:
            logging.getLogger().removeHandler(handler)
            self._elk_handler = None

    # ── ES thin client ────────────────────────────────────────────────────────

    @staticmethod
    def _build_es_client(cfg: ElasticsearchConfig) -> "_ESClient":
        return _ESClient(cfg)

    @staticmethod
    def _build_ls_writer(cfg: LogstashConfig) -> "_LogstashWriter":
        return _LogstashWriter(cfg)


class _ESClient:
    """
    Minimal Elasticsearch client using stdlib urllib — mirrors gateway's
    esClient struct.  No third-party elasticsearch-py dependency required.
    """

    def __init__(self, cfg: ElasticsearchConfig) -> None:
        self._cfg = cfg
        self._auth_header = self._build_auth(cfg)
        ssl_ctx = None
        if any([cfg.tls_ca_cert, cfg.tls_client_cert, cfg.insecure_skip_verify]):
            ssl_ctx = _build_ssl_context(
                cfg.tls_ca_cert, cfg.tls_client_cert, cfg.tls_client_key,
                cfg.insecure_skip_verify,
            )
        self._ssl_ctx = ssl_ctx

    @staticmethod
    def _build_auth(cfg: ElasticsearchConfig) -> str:
        if cfg.api_key:
            return f"ApiKey {cfg.api_key}"
        if cfg.bearer_token:
            return f"Bearer {cfg.bearer_token}"
        if cfg.username:
            import base64
            token = base64.b64encode(f"{cfg.username}:{cfg.password}".encode()).decode()
            return f"Basic {token}"
        return ""

    def ping(self) -> None:
        import urllib.error
        import urllib.request
        if not self._cfg.endpoints:
            raise RuntimeError("no elasticsearch endpoints configured")
        url = self._cfg.endpoints[0].rstrip("/") + "/_cluster/health"
        req = urllib.request.Request(url)
        if self._auth_header:
            req.add_header("Authorization", self._auth_header)
        try:
            with urllib.request.urlopen(req, context=self._ssl_ctx, timeout=5):
                pass
        except urllib.error.HTTPError as exc:
            if exc.code >= 500:
                raise RuntimeError(f"elasticsearch cluster health: HTTP {exc.code}") from exc
        except Exception as exc:
            raise RuntimeError(f"elasticsearch unreachable: {exc}") from exc

    def bulk_index(self, records: list[dict]) -> None:
        import json as _json
        import urllib.request
        if not records or not self._cfg.endpoints:
            return

        lines = []
        for rec in records:
            lines.append(_json.dumps({"index": {"_index": self._cfg.index}}))
            lines.append(_json.dumps(rec))
        body = "\n".join(lines).encode() + b"\n"

        url = self._cfg.endpoints[0].rstrip("/") + "/_bulk"
        if self._cfg.pipeline:
            url += f"?pipeline={self._cfg.pipeline}"

        req = urllib.request.Request(url, data=body, method="POST")
        req.add_header("Content-Type", "application/x-ndjson")
        if self._auth_header:
            req.add_header("Authorization", self._auth_header)

        try:
            with urllib.request.urlopen(req, context=self._ssl_ctx, timeout=30) as resp:
                if resp.status >= 300:
                    raise RuntimeError(f"elasticsearch bulk returned HTTP {resp.status}")
        except Exception:
            logger.warning("elk: ES bulk_index failed", exc_info=True)


class _LogstashWriter:
    """
    Logstash HTTP input writer — mirrors gateway's logstashWriter struct.
    Sends NDJSON batches to the Logstash HTTP input plugin.
    """

    def __init__(self, cfg: LogstashConfig) -> None:
        self._cfg = cfg
        self._ssl_ctx = None
        if cfg.use_tls or cfg.tls_ca_cert:
            self._ssl_ctx = _build_ssl_context(
                cfg.tls_ca_cert, cfg.tls_client_cert, cfg.tls_client_key,
                cfg.insecure_skip_verify,
            )

    def _url(self) -> str:
        scheme = "https" if self._cfg.use_tls else "http"
        return f"{scheme}://{self._cfg.endpoint}"

    def ping(self) -> None:
        import urllib.request
        req = urllib.request.Request(self._url(), method="GET")
        try:
            with urllib.request.urlopen(req, context=self._ssl_ctx,
                                         timeout=self._cfg.timeout_seconds):
                pass
        except Exception as exc:
            raise RuntimeError(f"logstash unreachable: {exc}") from exc

    def write_batch(self, records: list[dict]) -> None:
        import json as _json
        import urllib.request
        if not records:
            return
        body = ("\n".join(_json.dumps(r) for r in records) + "\n").encode()
        req  = urllib.request.Request(self._url(), data=body, method="POST")
        req.add_header("Content-Type", "application/x-ndjson")
        try:
            with urllib.request.urlopen(req, context=self._ssl_ctx,
                                         timeout=self._cfg.timeout_seconds) as resp:
                if resp.status >= 300:
                    raise RuntimeError(f"logstash returned HTTP {resp.status}")
        except Exception:
            logger.warning("elk: Logstash write_batch failed", exc_info=True)

File: gateway-service/internal/models/monitoring/elk.go
package monitoring

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"strings"
	"sync"
	"time"

	monpb "gateway-service/internal/gen/proto/go/vartrack/v1/models/monitoring"

	"gateway-service/internal/protoutil"
)

// ELKBackend ships structured log records to Elasticsearch (Bulk API) or
// via a Logstash HTTP input.
//
// It installs itself as a second slog handler via a multiHandler so every
// slog.Info/Warn/Error/Debug call in the application is automatically captured
// and shipped — no call-site changes required.
//
// Log flow:
//
//	slog.Info("...")
//	  → multiHandler
//	       ├─ original JSON-stdout handler  (preserved)
//	       └─ elkHandler → enqueue → flushWorker → ES Bulk API / Logstash
//
// Implements: Backend (Name / Ping / Shutdown)
type ELKBackend struct {
	cfg             *monpb.ELKConfig
	es              *esClient
	ls              *logstashWriter
	buf             []json.RawMessage
	bufMu           sync.Mutex
	flushCh         chan struct{}
	stopCh          chan struct{}
	wg              sync.WaitGroup
	previousHandler slog.Handler // restored on Shutdown
}

// NewELKBackend wires the ELK log pipeline and installs the slog multi-handler.
// Returns a disabled no-op backend (not an error) when enabled: false.
func NewELKBackend(ctx context.Context, cfg *monpb.ELKConfig) (*ELKBackend, error) {
	b := &ELKBackend{
		cfg:     cfg,
		flushCh: make(chan struct{}, 1),
		stopCh:  make(chan struct{}),
	}

	if !cfg.GetEnabled() {
		slog.Info("elk: disabled")
		return b, nil
	}

	esCfg := cfg.GetElasticsearch()
	if esCfg == nil {
		return nil, fmt.Errorf("elk: elasticsearch config is required")
	}

	es, err := newESClient(esCfg)
	if err != nil {
		return nil, fmt.Errorf("elk: build ES client: %w", err)
	}
	b.es = es

	if lsCfg := cfg.GetLogstash(); lsCfg != nil {
		ls, err := newLogstashWriter(lsCfg)
		if err != nil {
			return nil, fmt.Errorf("elk: build Logstash writer: %w", err)
		}
		b.ls = ls
	}

	// Capture current handler and fan out to both old + ELK.
	b.previousHandler = slog.Default().Handler()
	slog.SetDefault(slog.New(newMultiHandler(b.previousHandler, newELKHandler(b, cfg))))

	b.wg.Add(1)
	go b.flushWorker()

	slog.Info("elk: started",
		"service", cfg.GetServiceName(),
		"index", esCfg.GetIndex(),
		"via_logstash", cfg.GetLogstash() != nil,
	)
	return b, nil
}

func (b *ELKBackend) Name() string { return "elk" }

// Ping checks Elasticsearch (and Logstash if configured) via HTTP.
func (b *ELKBackend) Ping(ctx context.Context) error {
	if b.es == nil {
		return nil
	}
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()

	if err := b.es.ping(ctx); err != nil {
		return fmt.Errorf("elk ping (elasticsearch): %w", err)
	}
	if b.ls != nil {
		if err := b.ls.ping(ctx); err != nil {
			return fmt.Errorf("elk ping (logstash): %w", err)
		}
	}
	return nil
}

// Shutdown stops the flush worker, flushes remaining records, then restores
// the original slog handler so post-shutdown logs still appear on stdout.
func (b *ELKBackend) Shutdown(ctx context.Context) error {
	if b.stopCh != nil {
		select {
		case <-b.stopCh:
		default:
			close(b.stopCh)
		}
	}
	b.wg.Wait()

	if err := b.flush(ctx); err != nil {
		slog.Warn("elk: final flush error", "error", err)
	}

	if b.previousHandler != nil {
		slog.SetDefault(slog.New(b.previousHandler))
	}

	slog.Info("elk: shut down")
	return nil
}

// enqueue is called by elkHandler.Handle for every captured log record.
func (b *ELKBackend) enqueue(record json.RawMessage) {
	maxDocs := int(b.cfg.GetElasticsearch().GetBulkMaxDocs())
	if maxDocs == 0 {
		maxDocs = 500
	}

	b.bufMu.Lock()
	b.buf = append(b.buf, record)
	shouldFlush := len(b.buf) >= maxDocs
	b.bufMu.Unlock()

	if shouldFlush {
		select {
		case b.flushCh <- struct{}{}:
		default:
		}
	}
}

func (b *ELKBackend) flushWorker() {
	defer b.wg.Done()

	interval := protoutil.DurationOrDefault(b.cfg.GetElasticsearch().GetFlushInterval(), 5*time.Second)
	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-b.stopCh:
			return
		case <-b.flushCh:
			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
			_ = b.flush(ctx)
			cancel()
		case <-ticker.C:
			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
			_ = b.flush(ctx)
			cancel()
		}
	}
}

func (b *ELKBackend) flush(ctx context.Context) error {
	b.bufMu.Lock()
	if len(b.buf) == 0 {
		b.bufMu.Unlock()
		return nil
	}
	batch := b.buf
	b.buf = nil
	b.bufMu.Unlock()

	if b.ls != nil {
		return b.ls.writeBatch(ctx, batch)
	}
	return b.es.bulkIndex(ctx, batch)
}

// ── slog handler ──────────────────────────────────────────────────────────────

type elkHandler struct {
	backend *ELKBackend
	cfg     *monpb.ELKConfig
	attrs   []slog.Attr
	group   string
}

func newELKHandler(backend *ELKBackend, cfg *monpb.ELKConfig) *elkHandler {
	return &elkHandler{backend: backend, cfg: cfg}
}

func (h *elkHandler) Enabled(_ context.Context, _ slog.Level) bool { return true }

func (h *elkHandler) Handle(_ context.Context, r slog.Record) error {
	doc := map[string]any{
		"@timestamp":      r.Time.UTC().Format(time.RFC3339Nano),
		"level":           r.Level.String(),
		"message":         r.Message,
		"service.name":    h.cfg.GetServiceName(),
		"service.version": h.cfg.GetServiceVersion(),
		"environment":     h.cfg.GetEnvironment(),
	}

	// Static extra fields from ES config.
	if esCfg := h.cfg.GetElasticsearch(); esCfg != nil {
		for k, v := range esCfg.GetExtraFields() {
			doc[k] = v
		}
	}

	// Handler-level attributes (added via slog.With / WithAttrs).
	for _, a := range h.attrs {
		doc[h.key(a.Key)] = a.Value.Any()
	}

	// Record-level attributes.
	r.Attrs(func(a slog.Attr) bool {
		doc[h.key(a.Key)] = a.Value.Any()
		return true
	})

	b, err := json.Marshal(doc)
	if err != nil {
		return err
	}
	h.backend.enqueue(b)
	return nil
}

func (h *elkHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	newAttrs := make([]slog.Attr, len(h.attrs)+len(attrs))
	copy(newAttrs, h.attrs)
	copy(newAttrs[len(h.attrs):], attrs)
	return &elkHandler{backend: h.backend, cfg: h.cfg, attrs: newAttrs, group: h.group}
}

func (h *elkHandler) WithGroup(name string) slog.Handler {
	return &elkHandler{backend: h.backend, cfg: h.cfg, attrs: h.attrs, group: name}
}

func (h *elkHandler) key(k string) string {
	if h.group != "" {
		return h.group + "." + k
	}
	return k
}

// ── multiHandler ──────────────────────────────────────────────────────────────

type multiHandler struct{ handlers []slog.Handler }

func newMultiHandler(handlers ...slog.Handler) *multiHandler {
	return &multiHandler{handlers: handlers}
}

func (m *multiHandler) Enabled(ctx context.Context, level slog.Level) bool {
	for _, h := range m.handlers {
		if h.Enabled(ctx, level) {
			return true
		}
	}
	return false
}

func (m *multiHandler) Handle(ctx context.Context, r slog.Record) error {
	for _, h := range m.handlers {
		if h.Enabled(ctx, r.Level) {
			if err := h.Handle(ctx, r.Clone()); err != nil {
				return err
			}
		}
	}
	return nil
}

func (m *multiHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	out := make([]slog.Handler, len(m.handlers))
	for i, h := range m.handlers {
		out[i] = h.WithAttrs(attrs)
	}
	return &multiHandler{handlers: out}
}

func (m *multiHandler) WithGroup(name string) slog.Handler {
	out := make([]slog.Handler, len(m.handlers))
	for i, h := range m.handlers {
		out[i] = h.WithGroup(name)
	}
	return &multiHandler{handlers: out}
}

// ── Elasticsearch thin client ─────────────────────────────────────────────────

type esClient struct {
	http       *http.Client
	endpoints  []string
	index      string
	authHeader string
	pipeline   string
}

func newESClient(cfg *monpb.ElasticsearchConfig) (*esClient, error) {
	tlsCfg, err := buildBackendTLS(
		cfg.GetTlsCaCert(), cfg.GetTlsClientCert(), cfg.GetTlsClientKey(),
		cfg.GetInsecureSkipVerify(),
	)
	if err != nil {
		return nil, err
	}

	authHeader := ""
	switch {
	case cfg.GetApiKey() != "":
		authHeader = "ApiKey " + cfg.GetApiKey()
	case cfg.GetBearerToken() != "":
		authHeader = "Bearer " + cfg.GetBearerToken()
	case cfg.GetUsername() != "":
		req, _ := http.NewRequest(http.MethodGet, "", nil)
		req.SetBasicAuth(cfg.GetUsername(), cfg.GetPassword())
		authHeader = req.Header.Get("Authorization")
	}

	return &esClient{
		http: &http.Client{
			Transport: &http.Transport{TLSClientConfig: tlsCfg},
			Timeout:   30 * time.Second,
		},
		endpoints:  cfg.GetEndpoints(),
		index:      cfg.GetIndex(),
		authHeader: authHeader,
		pipeline:   cfg.GetPipeline(),
	}, nil
}

func (c *esClient) ping(ctx context.Context) error {
	if len(c.endpoints) == 0 {
		return fmt.Errorf("no elasticsearch endpoints configured")
	}
	url := strings.TrimSuffix(c.endpoints[0], "/") + "/_cluster/health"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	if c.authHeader != "" {
		req.Header.Set("Authorization", c.authHeader)
	}
	resp, err := c.http.Do(req)
	if err != nil {
		return fmt.Errorf("elasticsearch unreachable: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 500 {
		return fmt.Errorf("elasticsearch cluster health: %s", resp.Status)
	}
	return nil
}

func (c *esClient) bulkIndex(ctx context.Context, records []json.RawMessage) error {
	if len(records) == 0 {
		return nil
	}

	var buf bytes.Buffer
	for _, rec := range records {
		// Each Bulk API action requires an action meta line followed by the document.
		meta := map[string]any{
			"index": map[string]any{"_index": c.index},
		}
		if err := json.NewEncoder(&buf).Encode(meta); err != nil {
			return err
		}
		buf.Write(rec)
		buf.WriteByte('\n')
	}

	url := strings.TrimSuffix(c.endpoints[0], "/") + "/_bulk"
	if c.pipeline != "" {
		url += "?pipeline=" + c.pipeline
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, &buf)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/x-ndjson")
	if c.authHeader != "" {
		req.Header.Set("Authorization", c.authHeader)
	}

	resp, err := c.http.Do(req)
	if err != nil {
		return fmt.Errorf("elasticsearch bulk: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return fmt.Errorf("elasticsearch bulk returned %s", resp.Status)
	}
	return nil
}

// ── Logstash HTTP writer ──────────────────────────────────────────────────────

type logstashWriter struct {
	http *http.Client
	cfg  *monpb.LogstashConfig
}

func newLogstashWriter(cfg *monpb.LogstashConfig) (*logstashWriter, error) {
	tlsCfg, err := buildBackendTLS(
		cfg.GetTlsCaCert(), cfg.GetTlsClientCert(), cfg.GetTlsClientKey(),
		cfg.GetInsecureSkipVerify(),
	)
	if err != nil {
		return nil, err
	}
	timeout := protoutil.DurationOrDefault(cfg.GetTimeout(), 10*time.Second)
	return &logstashWriter{
		cfg: cfg,
		http: &http.Client{
			Transport: &http.Transport{TLSClientConfig: tlsCfg},
			Timeout:   timeout,
		},
	}, nil
}

func (lw *logstashWriter) ping(ctx context.Context) error {
	scheme := "http"
	if lw.cfg.GetUseTls() {
		scheme = "https"
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet,
		scheme+"://"+lw.cfg.GetEndpoint(), nil)
	if err != nil {
		return err
	}
	resp, err := lw.http.Do(req)
	if err != nil {
		return fmt.Errorf("logstash unreachable: %w", err)
	}
	defer resp.Body.Close()
	return nil
}

func (lw *logstashWriter) writeBatch(ctx context.Context, records []json.RawMessage) error {
	var buf bytes.Buffer
	for _, rec := range records {
		buf.Write(rec)
		buf.WriteByte('\n')
	}

	scheme := "http"
	if lw.cfg.GetUseTls() {
		scheme = "https"
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		scheme+"://"+lw.cfg.GetEndpoint(), &buf)
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/x-ndjson")

	resp, err := lw.http.Do(req)
	if err != nil {
		return fmt.Errorf("logstash write: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		return fmt.Errorf("logstash returned %s", resp.Status)
	}
	return nil
}

func (lw *logstashWriter) close() {}
